src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/csip/CSIPFactorSynchronisationService.kt

```diff
@@ -0,0 +1,157 @@
+  private val mappingApiService: CSIPMappingService,
+  private val csipService: CSIPService,
+  private val telemetryClient: TelemetryClient,
+  private val queueService: SynchronisationQueueService,
+) {
+  private companion object {
+    val log: Logger = LoggerFactory.getLogger(this::class.java)
+  }
+
+  suspend fun csipFactorInserted(event: CSIPFactorEvent) {
+    val telemetry =
+      mutableMapOf(
+        "nomisCSIPReportId" to event.csipReportId,
+        "offenderNo" to event.offenderIdDisplay,
+        "nomisCSIPFactorId" to event.csipFactorId,
+      )
+
+    if (event.auditModuleName == "DPS_SYNCHRONISATION") {
+      telemetryClient.trackEvent("csip-factor-synchronisation-skipped", telemetry)
+      return
+    }
+
+    val nomisFactorResponse = nomisApiService.getCSIPFactor(event.csipFactorId)
+    // Get the report mapping
+    mappingApiService.findCSIPReportByNomisId(nomisCSIPReportId = event.csipReportId)
+      ?.let { reportMapping ->
+        mappingApiService.findCSIPFactorByNomisId(nomisCSIPFactorId = event.csipFactorId)
+          ?.let {
+            // Should never happen - factor should not exist in the mapping table
+            telemetryClient.trackEvent(
+              "csip-factor-synchronisation-created-ignored",
+              telemetry + ("dpsCSIPFactorId" to it.dpsCSIPFactorId) + ("reason" to "CSIP Factor already mapped"),
+            )
+          }
+          ?: run {
+            // HAPPY PATH
+            csipService.createCSIPFactor(
+              reportMapping.dpsCSIPId,
+              nomisFactorResponse.toDPSFactorRequest(),
+              nomisFactorResponse.createdBy,
+            ).also {
+              telemetry.put("dpsCSIPFactorId", it.factorUuid.toString())
+
+              tryToCreateCSIPFactorMapping(nomisFactorResponse, it, telemetry)
+                .also { mappingCreateResult ->
+                  if (mappingCreateResult == MappingResponse.MAPPING_FAILED) telemetry.put("mapping", "initial-failure")
+                  telemetryClient.trackEvent(
+                    "csip-factor-synchronisation-created-success",
+                    telemetry,
+                  )
+                }
+            }
+          }
+      }
+      ?: run {
+        // Should never happen - report should exist in the mapping table
+        telemetryClient.trackEvent(
+          "csip-factor-synchronisation-created-failed",
+          telemetry + ("reason" to "CSIP Report for CSIP factor not mapped"),
+        )
+        throw IllegalStateException("Received CSIP_FACTORS_INSERTED for csip Report that has never been created/mapped")
+      }
+  }
+
+  suspend fun csipFactorDeleted(event: CSIPFactorEvent) {
+    val telemetry =
+      mutableMapOf(
+        "nomisCSIPReportId" to event.csipReportId,
+        "offenderNo" to event.offenderIdDisplay,
+        "nomisCSIPFactorId" to event.csipFactorId,
+      )
+
+    if (event.auditModuleName == "DPS_SYNCHRONISATION") {
```

Comment:
not sure this works for delete records since the audit module would be from the last update not the actual delete